linq.d.ts

```diff
@@ -142,14 +143,20 @@ declare namespace Enumerable {
     sum(selector?: (element: T) => number): number;
     elementAt(index: number): T;
     elementAtOrDefault(index: number, defaultValue?: T): T | undefined;
+    first<TOther extends T>(predicate?: (element: T, index: number) => element is TOther): TOther;
     first(predicate?: (element: T, index: number) => boolean): T;
-    firstOrDefault(predicate?: (element: T, index: number) => boolean, defaultValue?: T): T | undefined;
-    firstOrDefault(defaultValue?: T): T | undefined;
+    firstOrDefault<TOther extends T, TDefault>(predicate?: (element: T, index: number) => element is TOther, defaultValue?: TDefault): TOther | TDefault | undefined;
+    firstOrDefault<TDefault>(predicate?: (element: T, index: number) => boolean, defaultValue?: TDefault): T | TDefault | undefined;
+    firstOrDefault<TDefault>(defaultValue?: TDefault): T | TDefault | undefined;
+    last<TOther extends T>(predicate?: (element: T, index: number) => element is TOther): TOther;
     last(predicate?: (element: T, index: number) => boolean): T;
-    lastOrDefault(predicate?: (element: T, index: number) => boolean, defaultValue?: T): T | undefined;
-    lastOrDefault(defaultValue?: T): T | undefined;
+    lastOrDefault<TOther extends T, TDefault>(predicate?: (element: T, index: number) => element is TOther, defaultValue?: TDefault): TOther | TDefault | undefined;
+    lastOrDefault<TDefault>(predicate?: (element: T, index: number) => boolean, defaultValue?: TDefault): T | TDefault | undefined;
+    lastOrDefault<TDefault>(defaultValue?: TDefault): T | TDefault | undefined;
+    single<TOther extends T>(predicate?: (element: T, index: number) => element is TOther): TOther;
     single(predicate?: (element: T, index: number) => boolean): T;
-    singleOrDefault(predicate?: (element: T, index: number) => boolean, defaultValue?: T): T | undefined;
+    singleOrDefault<TOther extends T, TDefault>(predicate?: (element: T, index: number) => element is TOther, defaultValue?: TDefault): TOther | TDefault | undefined;
+    singleOrDefault<TDefault>(predicate?: (element: T, index: number) => boolean, defaultValue?: TDefault): T | TDefault | undefined;
```

Comment:
There's a fly-by improvement for the `...OrDefault` functions here. It is common to do something like the code below, which uses a different type for the default value than the type of the elements of the input array.
```ts
const input = [23, 42, -1, 452]; // `input` is a `number[]`
const output = Enumerable
    .from(input)
    .firstOrDefault(value => value > 0, null); // prior to this commit, this would not compile because `null` is not a `number`
```

As I am adding support for type guard predicates, this came to the forefront as an issue. If I were to follow the pattern of the pre-existing functions, then the new overload would be as follows. Note that the return type is `T | TOther | undefined`, which allows for the predicate to work but still confuses the point by insisting that the default has to be a `T`. By adding the additional generic type, the default value is allowed to be any type appropriate for the call site, whether that be a `T`, a `TOther`, a `null` or anything else the developer uses as the default value.
```ts
firstOrDefault<TOther extends T>(predicate?: (element: T, index: number) => element is TOther, defaultValue?: T): T | TOther | undefined;
```